File: src/main/java/no/nav/syfo/config/VaultHikariConfig.java
package no.nav.syfo.config;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import lombok.extern.slf4j.Slf4j;
import no.nav.vault.jdbc.hikaricp.HikariCPVaultUtil;
import no.nav.vault.jdbc.hikaricp.VaultError;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;

@Slf4j
@EnableJpaRepositories("no.nav.syfo.repository")
public class VaultHikariConfig {

    @Value("${vault.enabled:true}")
    public boolean enabled = true;
    @Value("${vault.backend}")
    public String databaseBackend;
    @Value("${vault.role}")
    public String databaseRole;
    @Value("${vault.admin}")
    public String databaseAdminrole;

    @Bean
    public HikariDataSource dataSource(DataSourceProperties properties) throws VaultError {
        HikariConfig config = createHikariConfig(properties);
        if (enabled) {
            return HikariCPVaultUtil.createHikariDataSourceWithVaultIntegration(config, databaseBackend, databaseRole);
        }
        config.setUsername("tullball");
        config.setPassword(properties.getPassword());
        return new HikariDataSource(config);
    }

    static HikariConfig createHikariConfig(DataSourceProperties properties) {
        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(properties.getUrl());
        config.setMinimumIdle(1);
        config.setMaximumPoolSize(2);
        return config;
    }

}

File: src/main/kotlin/no/nav/syfo/config/VaultHikariConfig.kt
package no.nav.syfo.config

import no.nav.vault.jdbc.hikaricp.HikariCPVaultUtil


class VaultHikariConfig (private var enabled: Boolean,
                         private var databaseBackend: String?,
                         private var databaseRole: String?,
                         private var databaseAdminrole: String?){

 /*   fun dataSource(properties: DataSourceProperties): HikariDataSource {
        return buildDatasource(properties)
    }

    @SneakyThrows
    fun buildDatasource(properties: DataSourceProperties): HikariDataSource {
        val config: HikariConfig = createHikariConfig(properties)
          val config = HikariConfig()
            config.setJdbcUrl(properties.getUrl())
            config.setMinimumIdle(1)
            config.setMaximumPoolSize(2)
        if (enabled) {
            return HikariCPVaultUtil.createHikariDataSourceWithVaultIntegration(
                config,
                databaseBackend,
                databaseAdminrole
            )
        }
        config.setUsername(properties.getUsername())
        config.setPassword(properties.getPassword())
        return HikariDataSource(config)
    }

    @Bean
    fun flywayMigrationStrategy(properties: DataSourceProperties): FlywayMigrationStrategy {
        return if (enabled) {
            FlywayMigrationStrategy { flyway ->
                Flyway.configure()
                    .dataSource(buildDatasource(properties))
                    .initSql(String.format("SET ROLE \"%s\"", databaseAdminrole))
                    .load()
                    .migrate()
            }
        } else FlywayMigrationStrategy { flyway ->
            Flyway.configure()
                .dataSource(buildDatasource(properties))
                .load()
                .migrate()
        }
        // Ikke kjør rolle settingen i testing. De feiler
    }

    companion object {
        fun createHikariConfig(properties: DataSourceProperties): HikariConfig {
            val config = HikariConfig()
            config.setJdbcUrl(properties.getUrl())
            config.setMinimumIdle(1)
            config.setMaximumPoolSize(2)
            return config
        }
    }*/
}
